utils/vfs/vfs.cpp

```diff
@@ -0,0 +1,120 @@
+#include <algorithm>
+#include <fuzzuf/exceptions.hpp>
+#include <fuzzuf/utils/filesystem.hpp>
+#include <fuzzuf/utils/vfs/vfs.hpp>
+namespace fuzzuf::utils::vfs {
+
+VFS::VFS(std::vector<fs::path> &&allowed_path_) {
+  if (allowed_path_.empty()) {
+    allowed_path.reset(new std::vector<fs::path>{});
+    current_workdir = "/nowhere";
+    return;
+  }
+  std::vector<fs::path> duplicated_allowed_path;
+  duplicated_allowed_path.reserve(allowed_path_.size());
+  if (!allowed_path_[0].has_root_directory())
+    throw exceptions::invalid_argument(
+        "First allowed path must contain root directory", __FILE__, __LINE__);
+  {
+    auto can = allowed_path_[0].lexically_normal();
+    current_workdir = can;
+  }
+  std::transform(allowed_path_.begin(), allowed_path_.end(),
+                 std::back_inserter(duplicated_allowed_path),
+                 [this](const auto &v) {
+                   auto abs = Absolute(v);
+                   auto can = abs.lexically_normal();
+                   return can;
+                 });
+  std::sort(duplicated_allowed_path.begin(), duplicated_allowed_path.end());
+  allowed_path.reset(new std::vector<fs::path>{});
+  allowed_path->push_back(std::move(duplicated_allowed_path[0]));
+  current_workdir = (*allowed_path)[0];
+  fs::path prev = duplicated_allowed_path[0];
+  auto prev_len = std::distance(prev.begin(), prev.end());
+  /*
+   * Remove duplicated or included paths.
+   */
+  for (auto cur = std::next(duplicated_allowed_path.begin());
+       cur != duplicated_allowed_path.end(); ++cur) {
+    auto cur_len = std::distance(cur->begin(), cur->end());
+    /*
+     * Since the paths are sorted, the path shorter than previous one definitely
+     * means no inclusion nor duplication. So the path should be left.
+     */
+    if (prev_len > cur_len) {
+      prev = *cur;
+      allowed_path->push_back(std::move(*cur));
+      prev_len = std::distance(prev.begin(), prev.end());
+    }
+    /*
+     * If the path doesn't contain previous path as prefix, the path is not
+     * included nor duplicated. So the path should be left.
+     */
+    else if (!std::equal(prev.begin(), prev.end(), cur->begin(),
+                         std::next(cur->begin(), prev_len))) {
+      prev = *cur;
+      allowed_path->push_back(std::move(*cur));
+      prev_len = std::distance(prev.begin(), prev.end());
+    }
+    /*
+     * Otherwise, the path is same as or included by previous one.
+     * So the path should be discarded.
+     */
```

Comment:
These comments are very good. 
I think this part can be improved further by introducing a bool variable(which may be excessive though).

```suggestion
    bool should_retain;
    /*
     * Since the paths are sorted, the path shorter than previous one definitely
     * means no inclusion nor duplication. So the path should be left.
     */
    if (prev_len > cur_len) {
        should_retain = true;
    }
    /*
     * If the path doesn't contain previous path as prefix, the path is not
     * included nor duplicated. So the path should be left.
     */
    else if (!std::equal(prev.begin(), prev.end(), cur->begin(),
                         std::next(cur->begin(), prev_len))) {
        should_retain = true;
    }
    /*
     * Otherwise, the path is same as or included by previous one.
     * So the path should be discarded.
     */
     else {
         should_retain = false;
     }
     
     if (should_retain) {
        prev = *cur;
        allowed_path->push_back(std::move(*cur));
        prev_len = std::distance(prev.begin(), prev.end());
     }
```